Print unpingable gateway disclaimer

giantpanda9 opened the pull request:
Print unpingable gateway disclaimer

- Changed file netconfig.cpp
   - Added new line to help text to
     bmc ifconfig gateway help
     directive output
   - the new help text will displayed
     only when the following command
     bmc ifconfig gateway help executed

End-User-Impact: To display a disclaimer in CLI part
Signed-off-by: Nikolay Chegodaev <n.chegodaev@yadro.com>

giantpanda9 commented:
I have limited NOTE: discliamer which is displayed in CLI with 80 characcter limit - this looks like as present on screenshot attached
![Limit80Char](https://user-images.githubusercontent.com/61800770/191509865-9421788b-fb3f-43ab-a83c-638008e319bd.png)

Please advise if it looks fine, or not?

giantpanda9 commented:
I have followed the instructions provided:
1) Cloned this repository to local machine
2) Created a branch here - moved changes a debugged in opembmc
3) Create a branch here and PR to github
4) Changed the line in recipe as instructed
5) The last thing I bump'ed the local openbmc repository
@nest1ing  Is this fine?

nest1ing commented:
> I have limited NOTE: discliamer which is displayed in CLI with 80 characcter limit - this looks like as present on screenshot attached ![Limit80Char](https://user-images.githubusercontent.com/61800770/191509865-9421788b-fb3f-43ab-a83c-638008e319bd.png)
> 
> Please advise if it looks fine, or not?

It looks fine for `bmc ifconfig gateway help` but weird for `bmc ifconfig help`:
```
  hostname   Set host name
             Command format: hostname NAME

  gateway    Set default gateway
NOTE: Since the system does not allow unpingable gateway address, make sure the 
gateway address is pingable.
             Command format: gateway IP

  ip         Add or remove static IP address (default mask: IPv4/24, IPv6/64)
             Command format: ip {INTERFACE} {add|del} IP[/MASK]
```
Cannot we show the such message only when the specified gateway was declined?

nest1ing commented:
> Solves of BBMC-1172

This is public project and the such reference to our internal resource is unnecessary.

nest1ing commented:
>        @nest1ing  Is this fine?

The PR looks fine.

nest1ing reviewed (changes requested):
As I explained above, I dislike how it looks at the help for all commands.
And the commit message should be reformatted.

giantpanda9 commented:
> > I have limited NOTE: discliamer which is displayed in CLI with 80 characcter limit - this looks like as present on screenshot attached ![Limit80Char](https://user-images.githubusercontent.com/61800770/191509865-9421788b-fb3f-43ab-a83c-638008e319bd.png)
> > Please advise if it looks fine, or not?
> 
> It looks fine for `bmc ifconfig gateway help` but weird for `bmc ifconfig help`:
> 
> ```
>   hostname   Set host name
>              Command format: hostname NAME
> 
>   gateway    Set default gateway
> NOTE: Since the system does not allow unpingable gateway address, make sure the 
> gateway address is pingable.
>              Command format: gateway IP
> 
>   ip         Add or remove static IP address (default mask: IPv4/24, IPv6/64)
>              Command format: ip {INTERFACE} {add|del} IP[/MASK]
> ```
> 
> Cannot we show the such message only when the specified gateway was declined?

@nest1ing Not quite sure if the case "checking if gateway is valid and if so to display a message" could be implemented as part of this task, but I can try to fix how the message is displayed for the general help mode - to move it below the main text, when the help text is displayed for bmc ifconfig help.

1) Should I move it below the whole help display?
2) Or Should I add paddings so that note still stay near the gateway item?

giantpanda9 commented:
@nest1ing 
I have changed the commit commentary formatting
I have fixed NOTE: disclaimer app

![BMC1172_NOTES_Changes](https://user-images.githubusercontent.com/61800770/191788387-9f70c4d8-9b67-47cc-923b-2e9f7b89c0cf.png)

earance, please check the screenshot to see if everything now correct in requested sections

@ikmsk10 Could you please check is treated the commit in accordance to the rules in company? - I am sorry, but the pipeline and merging policies are still a novation to me, I do no wish to corrupt anything...

nest1ing reviewed (changes requested):
Seems You've done something wrong via `git rebase --interactive`.
As a result this PR modifies the commit already applied to master.

giantpanda9 commented:
> Seems You've done something wrong via `git rebase --interactive`. As a result this PR modifies the commit already applied to master.

@nest1ing 

Did as you advised, commit seems fixed, bump also done without errors

ikmsk10 commented:
@nest1ing, 
@giantpanda9 tells me that you have discussion on the team chat about the way to define `details` field.
Your proposal looks a little bit strange to me, to define `details` in the `Command` struct. Yes, this field will has a default initialization value `nullptr`, but this means it must be placed at the end of a struct definition, after `Handler fn`. 
So, the initialization of Command item of the array will look a little bit strange.
```
 {"gateway", "IP", "Set default gateway", cmdGateway, "<Details>"},
```

I haven't objection against your solution at all, but seems to me the another one struct like `CommandDetails` looks a little more useful and pretty. What do you think about this?

giantpanda9 commented:
> @nest1ing, @giantpanda9 tells me that you have discussion on the team chat about the way to define `details` field. Your proposal looks a little bit strange to me, to define `details` in the `Command` struct. Yes, this field will has a default initialization value `nullptr`, but this means it must be placed at the end of a struct definition, after `Handler fn`. So, the initialization of Command item of the array will look a little bit strange.
> 
> ```
>  {"gateway", "IP", "Set default gateway", cmdGateway, "<Details>"},
> ```
> 
> I haven't objection against your solution at all, but seems to me the another one struct like `CommandDetails` looks a little more useful and pretty. What do you think about this?

- It seems like current solutions looks a little more systematic, there are not so much changes to the existent code, only additions - that could be considered as part of pro's of the current solutions. 

- The solution that @nest1ing proposed can make the code itself a little more compact (as we get rid of the additional loop) at the cost of the code being a little messy to read - strange new field after the function handler, that could be confusing for a user who will this code at the first time

giantpanda9 committed:
Print unpingable gateway disclaimer

- Changed file netconfig.cpp
   - Added new line to help text to
     bmc ifconfig gateway help
     directive output
   - the new help text will displayed
     only when the following command
     bmc ifconfig gateway help executed
End-User-Impact: To display a disclaimer in CLI part
Signed-off-by: Nikolay Chegodaev <n.chegodaev@yadro.com>

giantpanda9 commented:
> @nest1ing, @giantpanda9 tells me that you have discussion on the team chat about the way to define `details` field. Your proposal looks a little bit strange to me, to define `details` in the `Command` struct. Yes, this field will has a default initialization value `nullptr`, but this means it must be placed at the end of a struct definition, after `Handler fn`. So, the initialization of Command item of the array will look a little bit strange.
> 
> ```
>  {"gateway", "IP", "Set default gateway", cmdGateway, "<Details>"},
> ```
> 
> I haven't objection against your solution at all, but seems to me the another one struct like `CommandDetails` looks a little more useful and pretty. What do you think about this?

@nest1ing @ikmsk10  Fixed as discussed - the NOTE: text appearance logic is not change still in accordance to the screenshot below:

![Uploading BMC1172_NOTES_Changes.png…]()

Please re-review

nest1ing commented on src/netconfig.cpp:
```diff
@@ -379,7 +381,7 @@ static const Command ifconfigCommands[] = {
     {"mac", "{INTERFACE} MAC", "Set MAC address", cmdMac},
     {"hostname", "NAME", "Set host name", cmdHostname},
-    {"gateway", "IP", "Set default gateway", cmdGateway},
+    {"gateway", "IP", "Set default gateway", cmdGateway, "NOTE: The system doesn't allow unreachable gateways. Please ensure that your gateway \nbelongs to the network defined by the IP address and the Netmask."},
```
This message doesn't fit into 80 characters. I suggest to move `gateway` at the next line.
It will look like:
```
NOTE: The system doesn't allow unreachable gateways. Please ensure that your
gateway belongs to the network defined by the IP address and the netmask.
```
Capitalized `Netmask` seems to me strange. I would prefer to write it in lower case.

nest1ing replied:
Moreover, in the code this line also looks weird.
Could you reformat it to fit into 80 characters long?
I know that other lines also don't fit, but this line is too long. ;)

giantpanda9 commented on src/netconfig.cpp:
```diff
@@ -379,7 +381,7 @@ static const Command ifconfigCommands[] = {
     {"mac", "{INTERFACE} MAC", "Set MAC address", cmdMac},
     {"hostname", "NAME", "Set host name", cmdHostname},
-    {"gateway", "IP", "Set default gateway", cmdGateway},
+    {"gateway", "IP", "Set default gateway", cmdGateway, "NOTE: The system doesn't allow unreachable gateways. Please ensure that your \ngateway belongs to the network defined by the IP address and the netmask."},
```
@nest1ing @ikmsk10 Done as requested removed cepitalization from netmask and gateway moved into new line
![Resolved](https://user-images.githubusercontent.com/61800770/191998726-0962db9b-46a4-48ac-af6a-dd9aa9f508d6.png)

nest1ing replied:
I've seen. But in the source this line is still too long.

giantpanda9 replied:
> I've seen. But in the source this line is still too long.

@nest1ing Then I am confused a little, what do you wish me to do?
Do you want me not only move the "gateway"-starting part in the help text output to the new line, but also break the line in source code onto two lines?

Like so:
![BreakLine](https://user-images.githubusercontent.com/61800770/192058496-904b2061-7d8c-4ea5-bef3-7021d03cf6a2.png)

giantpanda9 replied:
@nest1ing 
- Fixed line breaking in netconfig.cpp source:
![BreakLine](https://user-images.githubusercontent.com/61800770/192232827-a9a956a8-f325-42bf-b11b-405454d6e4b1.png)
- Output still the same and not broken:
![NewHelp](https://user-images.githubusercontent.com/61800770/192232944-bb7bfd52-b0d5-4a16-8832-8e733c2bcc91.png)

Could you please review again?

AlexanderAmelkin commented:
This looks obsoleted by #24
